clang/include/indexstore/indexstore.h

```diff
@@ -100,12 +100,16 @@
 
 #if defined(__has_extension)
 #if __has_feature(objc_fixed_enum)
-# define INDEXSTORE_OPTIONS(_type, _name) enum INDEXSTORE_OPTIONS_ATTRS _name : _type _name; enum INDEXSTORE_OPTIONS_ATTRS _name : _type
+#ifdef __cplusplus
+# define INDEXSTORE_OPTIONS(_type, _name) enum INDEXSTORE_OPTIONS_ATTRS _name : _type
+#else
+# define INDEXSTORE_OPTIONS(_type, _name) enum _name : _type; typedef enum _name _name; enum INDEXSTORE_OPTIONS_ATTRS _name : _type
```

Comment:
This macro has been `typedef` + `enum` definition. With this change it's an `enum` declaration + `typedef` + `enum` definition. Is it invalid to create a `typedef` for an `enum` before it's defined?

Reply:
That is not true. There is a semicolon in the old macro and therefore the old macro is a typedef + enum *declaration* (not definition). The C++ compiler is more strict now about what is allowed in enum declarations and I'm trying to fix that.

Reply:
I may have used _definition_ incorrectly, what I was trying to ask about was the addition of the forward declaration.

Here's the result of partial macro expansion (I left `INDEXSTORE_OPTIONS_ATTRS` as is), newlines added after the semicolons, and the enum constants excluded. In the old macro, it was `typedef` and `enum` definition.

```c
typedef enum INDEXSTORE_OPTIONS_ATTRS indexstore_symbol_property_t : uint64_t indexstore_symbol_property_t;
enum INDEXSTORE_OPTIONS_ATTRS indexstore_symbol_property_t : uint64_t {
// ...
} indexstore_symbol_language_t;
```

```c
enum indexstore_symbol_property_t : uint64_t;
typedef enum indexstore_symbol_property_t indexstore_symbol_property_t;
enum INDEXSTORE_OPTIONS_ATTRS indexstore_symbol_property_t : uint64_t {
// ...
} indexstore_symbol_language_t;
```

With your change, there's an added forward declaration, and I was wondering if that's required. Compared to the previous implementation where a `typedef` was created for the `enum` that was not yet declared.

Reply:
Right. This was the best I could do to fix the error without throughly understanding the C++ parsing and semantic rules. As an aside, I think it’s really weird that this new warning defaults to error. It feels unprecedented but I don’t have the time to inquire on the clang development list.